timeline/src/Data/Timeline/Internal.hs

```diff
@@ -0,0 +1,299 @@
+{-# LANGUAGE ApplicativeDo #-}
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DeriveTraversable #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedLabels #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE ViewPatterns #-}
+
+module Data.Timeline.Internal
+  ( -- * Common
+    Timeline (..),
+    fromValues,
+    peek,
+    TimeRange (..),
+    isTimeAfterRange,
+    changes,
+
+    -- * effectiveFrom + optional effectiveTo
+    Record,
+    makeRecord,
+    recordEffectiveFrom,
+    recordEffectiveTo,
+    recordValue,
+    fromRecords,
+    Overlaps (..),
+    OverlapGroup (..),
+    unpackOverlapGroup,
+  )
+where
+
+import Data.Foldable.WithIndex (FoldableWithIndex (..))
+import Data.Functor.Contravariant (Contravariant, contramap)
+import Data.Functor.WithIndex (FunctorWithIndex (..))
+import Data.List (intercalate, sortOn)
+import Data.List.NonEmpty (NonEmpty ((:|)), nonEmpty)
+import Data.List.NonEmpty qualified as NonEmpty
+import Data.Map.Merge.Strict qualified as Map
+import Data.Map.Strict (Map)
+import Data.Map.Strict qualified as Map
+import Data.Maybe (mapMaybe, maybeToList)
+import Data.Semigroup.Foldable.Class (fold1)
+import Data.Set (Set)
+import Data.Time
+  ( UTCTime (..),
+  )
+import Data.Traversable.WithIndex (TraversableWithIndex (..))
+import GHC.Generics (Generic)
+import Language.Haskell.TH.Syntax qualified as TH (Lift (liftTyped))
+import Language.Haskell.TH.Syntax.Compat qualified as TH
+import Prelude
+
+-- | An infinite, discrete timeline for data type @a@.
+-- It always has a value for any time, but the set of values has a finite size.
+data Timeline a = Timeline
+  {- Internally this is represented by an initial value and a list of updated values.
+    Each change has an effective time and a new value.
+    This makes it possible to inspect the individual changes and to serialize the type,
+      but makes it impossible to represent timelines that change continuously. Continuous data
+      in computers' world is very rare, unless you are working with some math formulas.
+  -}
```

Comment:
Conal Elliott is crying in his cabin.

Also, should this be raised into the haddock?

Reply:
This needs to be revisited since the fields are public now. (They became public because we need to write a ToJSON in our private repository)